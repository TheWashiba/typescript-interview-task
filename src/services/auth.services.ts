import { API } from '~/constants';
import { getUrl } from '~/utils';

export const login = async (username: string, password: string) => {
  const url = getUrl(API.Login, {
    username,
    password,
  });

  const response = await fetch(url);
  const data = await response.json();
  const { token } = data;

  localStorage.setItem('token', token);
};

export const logout = async () => {
  try {
    await fetch(getUrl(API.Logout), {
      headers: {
        Authorization: `Bearer ${localStorage.getItem('token')}`,
      },
    });
  } catch (error) {
    throw new Error(error);
  } finally {
    localStorage.removeItem('token');
  }
};
