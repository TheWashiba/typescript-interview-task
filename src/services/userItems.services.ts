import { API } from '~/constants';
import { IUserItem } from '~/types';
import { getUrl } from '~/utils';

export const getUserItems = async (userId: string): Promise<IUserItem[]> => {
  const url = getUrl(API.Items, {
    userId,
  });

  const response = await fetch(url, {
    headers: {
      Authorization: `Bearer ${localStorage.getItem('token')}`,
    },
  });

  const data = await response.json();

  return data.items;
};

export const updateUserItem = async (item: IUserItem) =>
  await fetch(getUrl(API.Items), {
    method: 'POST',
    body: JSON.stringify(item),
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${localStorage.getItem('token')}`,
    },
  });
