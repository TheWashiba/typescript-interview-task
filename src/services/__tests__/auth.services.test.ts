import fetchMock from 'jest-fetch-mock';
import mockLocalStorage from '../../mocks/localStorage';
import { login } from '../auth.services';

describe('login service', () => {
  const credentials = {
    username: 'test',
    password: 'test',
  };

  beforeAll(() => {
    process.env.API_URL = 'test';
    fetchMock.enableMocks();
    mockLocalStorage();
  });

  afterEach(() => fetchMock.resetMocks());

  test('successful login sets token in localStorage', async () => {
    fetchMock.mockResponseOnce(JSON.stringify({ token: 'test-token' }));
    expect.assertions(1);

    await login(credentials.username, credentials.password);
    expect(localStorage.setItem).toHaveBeenCalledWith('token', 'test-token');
  });

  test('successful login sets token in localStorage', async () => {
    fetchMock.mockResponseOnce(JSON.stringify({ token: 'test-token' }));
    expect.assertions(1);

    await login(credentials.username, credentials.password);
    expect(fetch).toHaveBeenCalledWith(
      'test/api/login?password=test&username=test'
    );
  });

  test('failed login bubbles the error', async () => {
    fetchMock.mockReject(new Error('test'));
    expect.assertions(1);

    try {
      await login(credentials.username, credentials.password);
    } catch (error) {
      expect(error).toEqual(new Error('test'));
    }
  });
});
