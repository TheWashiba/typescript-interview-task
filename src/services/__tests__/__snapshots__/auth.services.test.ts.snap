// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logout logout sets auth header - "Bearer test-token" 1`] = `
Object {
  "calls": Array [
    Array [
      "test/api/logout",
      Object {
        "headers": Object {
          "Authorization": "Bearer test-token",
        },
      },
    ],
  ],
  "instances": Array [
    undefined,
  ],
  "invocationCallOrder": Array [
    9,
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
